insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt

```diff
@@ -0,0 +1,201 @@
+package dev.chrisbanes.accompanist.insets
+
+import android.os.Build
+import android.view.View
+import android.view.WindowInsets
+import androidx.annotation.RequiresApi
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.gesture.nestedscroll.NestedScrollConnection
+import androidx.compose.ui.gesture.nestedscroll.NestedScrollSource
+import androidx.compose.ui.platform.AmbientView
+import androidx.compose.ui.unit.Velocity
+import kotlin.math.roundToInt
+
+/**
+ * Remembers a [NestedScrollConnection] which scrolls the Android on-screen keyboard on/off
+ * screen as appropriate.
+ *
+ * @param scrollImeOffScreenWhenVisible Set to true to allow scrolling the IME off screen
+ * (from being visible), by an downwards scroll. Defaults to `true`.
+ * @param scrollImeOnScreenWhenNotVisible Set to true to allow scrolling the IME on screen
+ * (from not being visible), by an upwards scroll. Defaults to `true`.
+ */
+@ExperimentalAnimatedInsets
+@Composable
+fun rememberImeNestedScrollConnection(
+    scrollImeOffScreenWhenVisible: Boolean = true,
+    scrollImeOnScreenWhenNotVisible: Boolean = true,
+): NestedScrollConnection {
+    val view = AmbientView.current
+    return remember(view, scrollImeOffScreenWhenVisible, scrollImeOnScreenWhenNotVisible) {
+        ImeNestedScrollConnection(
+            view = view,
+            scrollImeOffScreenWhenVisible = scrollImeOffScreenWhenVisible,
+            scrollImeOnScreenWhenNotVisible = scrollImeOnScreenWhenNotVisible
+        )
+    }
+}
+
+/**
+ * A [NestedScrollConnection] which scrolls the Android on-screen keyboard on/off
+ * screen as appropriate, when the user scrolls content. This class may be made an internal
+ * library class in the future.
+ *
+ * You probably do not wish to use this directly, and should use
+ * [rememberImeNestedScrollConnection] instead.
+ *
+ * @param view The host Compose [View]. Usually this comes from [AmbientView].
+ * @param scrollImeOffScreenWhenVisible Set to true to allow scrolling the IME off screen
+ * (from being visible), by an downwards scroll. Defaults to `true`.
+ * @param scrollImeOnScreenWhenNotVisible Set to true to allow scrolling the IME on screen
+ * (from not being visible), by an upwards scroll. Defaults to `true`.
+ */
+@ExperimentalAnimatedInsets
+class ImeNestedScrollConnection(
+    private val view: View,
+    private val scrollImeOffScreenWhenVisible: Boolean,
+    private val scrollImeOnScreenWhenNotVisible: Boolean,
+) : NestedScrollConnection {
+
+    @delegate:RequiresApi(30)
+    private val imeAnimController by lazy(LazyThreadSafetyMode.NONE, ::SimpleImeAnimationController)
+
+    @get:RequiresApi(30)
+    private val imeVisible: Boolean
+        get() = view.rootWindowInsets.isVisible(WindowInsets.Type.ime())
+
+    override fun onPreScroll(available: Offset, source: NestedScrollSource): Offset {
+        if (Build.VERSION.SDK_INT < 30) {
```

Comment:
There are a number of API checks, could this class require API 30 to simplify it and move the check to when you create the ScrollConnection?

Reply:
I thought about that but went this way for 2 reasons:

1. It's hard to do an `if (SDK_INT < 30)` when using fluent Modifier chains.
2. We _may_ eventually get a version of `WindowInsetsAnimationController` which works < 30 (doubtful but you never know)

Reply:
Fair.